### PR TITLE
nativeimgutil: fix making a sparse file

### LINKED ISSUE #2713: vz: "Expanding to 100GiB" phase makes a non-sparse 100GiB image
>  I just found out that the built-in conversion needs more diskspace than `qemu-img convert`. While the end-result is still a sparse disk, it seems to require the full 100GB disk space temporarily, so you cannot convert from QCOW2 to RAW on a device with limited free space.
> 
> ```console
> $ df -h ~/.lima3
> Filesystem    Size    Used   Avail Capacity iused ifree %iused  Mounted on
> /dev/disk5    50Gi   692Mi    49Gi     2%      11  4.3G    0%   /Users/jan/.lima3
> 
> $ l start --vm-type vz
> ? Creating an instance "default" Proceed with the current configuration
> INFO[0001] Starting the instance "default" with VM driver "vz"
> …
> INFO[0002] Converting "/Users/jan/.lima3/default/basedisk" (qcow2) to a raw disk "/Users/jan/.lima3/default/diffdisk"
> 3.50 GiB / 3.50 GiB [-------------------------------------] 100.00% 206.87 MiB/s
> INFO[0019] Expanding to 100GiB
> FATA[0020] failed to convert "/Users/jan/.lima3/default/basedisk" to a raw disk "/Users/jan/.lima3/default/diffdisk": no space left on device
> ```
> 
> 
> Using `qemu-img convert` seems to require little extra space beyond what the new sparse file actually occupies.
> 

_Originally posted by @jandubois in https://github.com/lima-vm/lima/issues/2579#issuecomment-2403566144_

> > While the end-result is still a sparse disk
> 
> Actually, it is not, with the builtin conversion. It turns into a fully allocated disk. So this is even worse. That also might explain why it takes so long: it possibly writes the full 100GB to disk.
> 

_Originally posted by @jandubois in https://github.com/lima-vm/lima/issues/2579#issuecomment-2403587553_

- - -
https://github.com/lima-vm/lima/blob/bc774ded0a0710acb6716580719ace284d9ec3d4/pkg/nativeimgutil/nativeimgutil.go#L85-L91
https://github.com/lima-vm/lima/blob/bc774ded0a0710acb6716580719ace284d9ec3d4/pkg/nativeimgutil/nativeimgutil.go#L170-L175

### PR DESCRIPTION
**Pull request by AkihiroSuda**
"Expanding to 100GiB" phase was making a non-sparse 100GiB image

FIx #2713

### CODE REVIEW

File: pkg/nativeimgutil/nativeimgutil.go [hunk 1]
```diff
@@ -82,16 +81,24 @@ func ConvertToRaw(source, dest string, size *int64, allowSourceWithBackingFile b
...
 	// Resize
 	if size != nil {
+		// Reopen the closed fd
+		destTmpF, err = os.OpenFile(destTmp, os.O_RDWR, 0o600)
+		if err != nil {
+			return err
+		}
 		logrus.Infof("Expanding to %s", units.BytesSize(float64(*size)))
-		if err = MakeSparse(destTmpF, *size); err != nil {
+		if err = destTmpF.Truncate(*size); err != nil {
```
Comment (jandubois) [inline pkg/nativeimgutil/nativeimgutil.go:98]: Don't you still have to close the file again?

```suggestion
		if err = destTmpF.Truncate(*size); err != nil {
			_ = destTmpF.Close()
			return err
		}
		if err = destTmpF.Close(); err != nil {
			return err
		}
```

**Review by jandubois — DISMISSED** [review]
It looks good, except for the potential file handle leak.

I've tested it, and I could create a 5GB `fedora` VZ image in a 6GB RAM disk (on macOS aarch64).

File: pkg/nativeimgutil/nativeimgutil.go (hunk 1 above)
Comment (AkihiroSuda) [reply pkg/nativeimgutil/nativeimgutil.go:98]: done

**Comment by jandubois** [comment]
Sorry, I need to do more testing, but ran out of time for today. Will try again tomorrow.

It turns out that I was testing on a RAM disk formatted with HFS+, which does not support sparse files (thanks @nirs for pointing me at that possibility).

I thought I tested it again with this PR, and that it then worked, but I can no longer reproduce it. Need to retest all scenarios once more.

**Comment by jandubois** [comment]
> I've tested it, and I could create a 5GB `fedora` VZ image in a 6GB RAM disk (on macOS aarch64).

Ok, I was obviously confused for that one, as that is not a valid test of creating a sparse image.

**Comment by AkihiroSuda** [comment]
We may rather merge:
- #2718

**Pull request closed without merging**